### PR TITLE
Support Nerves TARGET_x environment variables

### PR DESCRIPTION
**Pull request by fhunleth**
The problem that this PR is trying to solve is that the architecture, os, abi detection is imperfect and can't be improved without some additional knowledge from Nerves. This affects several processors, but the easiest one to test is the Raspberry Pi 0. The architecture for the RPi0 is also "arm", so it looks the same even though it doesn't support the same instruction set as the RPi3 and other common 32-bit ARM processors. Without a CPU specification, zig generates ARMv7 instructions (this is a reasonable default for 2021). The RPi0 only supports ARMv6, so this doesn't work. 

The solution is that Nerves systems supply a set of environment variables. Nerves is currently using `TARGET_ABI`, `TARGET_OS`, `TARGET_ARCH`, and `TARGET_CPU` to make the options available to tools like `zigler`.

For example, these variables are set to the following for an RPi3:

```
TARGET_ABI=gnueabihf
TARGET_OS=linux
TARGET_ARCH=arm
TARGET_CPU=cortex_a53
```

For an RPi0, these are set as:

```
TARGET_ABI=gnueabihf
TARGET_OS=linux
TARGET_ARCH=arm
TARGET_CPU=arm1176jzf_s
```

Luckily, the conventions for specifying ABI, OS, ARCH, and CPU for `zig` seem to match other tools so nothing here is `zig`-specific.

The Nerves core team has added these variables to all of the official Nerves ports and quite a few unofficial ones. We've tested on 32-bit and 64-bit ARM boards and this seems to work well with the `hello_zig` example that you contributed and this PR.

I think this is working well enough to review and consider merging. The only issue is that `musl` is hardcoded since specifying `glibc` to `zig` didn't work. This doesn't seem to cause a problem `zig` programs that don't require , so I'm thinking that it could be addressed when we get to programs that actually care about this.

### CODE REVIEW

File: lib/zig/builder.ex [hunk 1]
```diff
@@ -119,6 +131,10 @@ defmodule Zig.Builder do
     end
   end
 
+  defp to_structdef(t = %{cpu: cpu}) do
+    ".{.default_target = .{.cpu_arch = .#{t.arch}, .os_tag = .#{t.os}, .abi = .#{t.abi}, .cpu_model = .{ .explicit = &std.Target.arm.cpu.#{cpu}}}}"
```
Comment (ityonemo) [inline lib/zig/builder.ex:135]: oh man that's amazing using the .explicit which gets thrown away when it's a non-arm arch.

**Comment by ityonemo** [comment]
I'm staying with a friend recovering from some surgery for a few weeks, but I'll probably be able to pick up my rpi from home later this week and will give it a whirl.

**Comment by fhunleth** [comment]
We released support in Nerves for setting the `TARGET_*` variables so I'm checking this out again. Everything builds fine. I've checked the options passed to Zig and they look good. At runtime on a Raspberry Pi Zero, I get

```
iex(9)> HelloZig.hello
** (RuntimeError) nif for function hello/0 not bound
    lib/hello_zig.ex:1: HelloZig.hello/0
```

I've tried a few things and I'm not getting any hints. Do you have any hints on how to debug this? No worries if not, I think that my next attempt is to annotate more code around the shared library loading since nothing appears to be returning errors.

**Comment by ityonemo** [comment]
so in theory there should be some logs that get emitted when the nif is bound, but they are set to level: debug.  If there's some way to have debug-level logging happen in nerves that would be the first place to start, it might also be useful to instrument logging into that magic pipeline.

This is where action is:
https://github.com/ityonemo/zigler/blob/master/lib/zig/compiler.ex#L78

And this is how it winds up getting called:
https://github.com/ityonemo/zigler/blob/master/lib/zig.ex#L255

**Comment by fhunleth** [comment]
FYI - The example is working on a Raspberry Pi Zero now. I had made a typo in the ABI spec and will have to make a new nerves_system_rpi0 release to fix that. However, I think the new code is working. Next chance I get, I'll look at this PR again and clean it up. 

Also, mix releases will remove the execute mode on the shared library created by zig. This causes Nerves to skip processing it as an executable which in turn caused the ABI mismatch to go unnoticed. It also results in the shared library being 900KB. Nerves normally strips shared libraries. In the example case, the shared library was around 250KB stripped. The fix to that is https://github.com/elixir-lang/elixir/pull/10709.  

I think this is finally really close!!!

**Comment by fhunleth** [comment]
@ityonemo I turned off the "draft" part of this PR since I think that it's ready to go. I've been testing against the official Nerves systems. Only the RPi Zero and original Model A and B don't work out of the box, but they will soon. Please let me know if you'd like me to change anything. 

I'm testing against https://github.com/nerves-project/nerves_examples/tree/hello_zig_updates/hello_zig which is basically your PR to `nerves-examples`, but with updates to the new Nerves libraries. I see that I need to update the README, but once this PR gets merged, I'll make a final pass on `hello_zig` and get it merged. Thanks again for your patience on this!!

**Review by ityonemo — APPROVED** [review]

**Pull request merged**